compiler-cli/src/build.rs

```diff
@@ -9,10 +9,10 @@ use crate::{
     build_lock::BuildLock,
     cli,
     dependencies::UseManifest,
-    fs::{self, ConsoleWarningEmitter},
+    fs::{self, ConsoleWarningEmitter, ProjectIO},
 };
 
-pub fn main(options: Options) -> Result<Package> {
+pub fn main(options: Options) -> Result<(ProjectCompiler<ProjectIO>, Package)> {
```

Comment:
Let's not expose the entire ProjectCompiler, it's a very large class and we don't need all that capability here.

Perhaps we can return something like this:

```rust
pub struct Built {
    pub root_package: Package,
    module_interfaces: im::HashMap<SmolStr, type_::Module> 
}
```

Then we can have a function for getting the main function for a given module

```rust
pub fn get_main_function(&self, module: &str) -> Result<StructWithTheInfo, Error> {
    todo!()
}
```

Generally we want to have as little in the IO providing `gleam_cli` as possible, with all the logic living in `gleam_core`, so this could live in `gleam_core`.

Unit tests for it would be fab also

Reply:
I did something similar to this. I thought it would be best to make `module_interfaces` public in that struct. I felt adding a `get_main_function` method to it would add some difficulty to reading the code for not much gain with encapsulation since I would need to access that has map outside of the compiler core crate anyway. I added a `get_function` method to `type_::Module` that is used to get the main function instead.

Since I haven't touched unit tests yet im not going to resolve this issue.

Reply:
I would like not to expose `module_interfaces`. It's an internal data structure that is going to change a lot so I want to keep it an internal implementation detail.

When would it need to be accessed?

Reply:
I was going to access it in the built struct. Now that i realize built should be internal as well what I was worrying about doesn't matter.

Reply:
ah right! Thank you 💜